crates/walrus-upload-relay/src/tip/config.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! The configuration for the Walrus Upload Relay's tipping system.
+
+use std::num::NonZeroU16;
+
+use serde::{Deserialize, Serialize};
+use sui_types::base_types::SuiAddress;
+use utoipa::ToSchema;
+use walrus_sdk::{
+    core::{EncodingType, encoding::encoded_blob_length_for_n_shards},
+    sui::SuiAddressSchema,
+};
+
+/// The kinds of tip that the proxy can choose to configure.
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+#[serde(rename_all = "snake_case")]
+#[schema(examples(
+        json!(TipKind::Const(31415)),
+        json!(TipKind::Linear{base: 101, encoded_size_mul_per_kb: 42})
+))]
+pub(crate) enum TipKind {
+    /// A constant tip.
+    Const(u64),
+    /// A tip that linearly depends on the encoded size of the blob.
+    Linear {
+        base: u64,
+        encoded_size_mul_per_kb: u64,
+    },
+}
+
+impl TipKind {
+    /// Returns the tip required for a blob of the given size, or `None` if no tip is required.
+    ///
+    /// Returns None if the blob size cannot be computed.
+    pub(crate) fn compute_tip(
+        &self,
+        n_shards: NonZeroU16,
+        unencoded_length: u64,
+        encoding_type: EncodingType,
+    ) -> Option<u64> {
+        Some(match self {
+            TipKind::Const(constant) => *constant,
+            TipKind::Linear {
+                base,
+                encoded_size_mul_per_kb,
+            } => {
+                base + encoded_blob_length_for_n_shards(n_shards, unencoded_length, encoding_type)?
+                    / 1024
+                    * encoded_size_mul_per_kb
```

Comment:
Question: Is it intentional that you round down here? In other similar cases (e.g., when computing storage cost) we use `div_ceil`.

Reply:
This will cause changes to the SDK and the docs. I'll try to address those today.